applications/main/nfc/plugins/supported_cards/bip.c

```diff
@@ -0,0 +1,372 @@
+        return t1->second - t2->second;
+    }
+    return 0;
+}
+
+static void print_bip_timestamp(const BipTimestamp* timestamp, FuriString* str) {
+    furi_assert(timestamp);
+    furi_assert(str);
+    furi_string_cat_printf(
+        str,
+        "%04u-%02u-%02u %02u:%02u:%02u",
+        timestamp->year,
+        timestamp->month,
+        timestamp->day,
+        timestamp->hour,
+        timestamp->minute,
+        timestamp->second);
+}
+
+static bool is_bip_block_empty(const MfClassicBlock* block) {
+    furi_assert(block);
+    // check if all but last byte are zero (last is checksum)
+    for(size_t i = 0; i < sizeof(block->data) - 1; i++) {
+        if(block->data[i] != 0) {
+            return false;
+        }
+    }
+    return true;
+}
+
+static void parse_uint16_le(const uint8_t* data, uint16_t* value) {
+    furi_assert(data);
+    furi_assert(value);
+
+    *value = (data[0]) | (data[1] << 8);
+}
+
+static void parse_uint32_le(const uint8_t* data, uint32_t* value) {
+    furi_assert(data);
+    furi_assert(value);
+
+    *value = (data[0]) | (data[1] << 8) | (data[2] << 16) | (data[3] << 24);
+}
+
+static void parse_uint16_txn_amount(const uint8_t* data, uint16_t* value) {
+    furi_assert(data);
+    furi_assert(value);
+
+    parse_uint16_le(data, value);
+    *value = *value >> 2;
+}
+
+typedef struct {
+    BipTimestamp timestamp;
+    uint16_t amount;
+} BipTransaction;
+
+static bool bip_parse(const NfcDevice* device, FuriString* parsed_data) {
+    furi_assert(device);
+    furi_assert(parsed_data);
+
+    bool parsed = true;
+
+    struct {
+        uint32_t card_id;
+        uint16_t balance;
+        uint16_t flags;
+        BipTimestamp trip_time_window;
+        BipTransaction top_ups[3];
+        BipTransaction charges[3];
+    } bip_data = {
+        .card_id = 0,
+        .balance = 0,
+        .flags = 0,
+        .trip_time_window = {0, 0, 0, 0, 0, 0},
+        .top_ups =
+            {
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+            },
+        .charges =
+            {
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+            },
+    };
+
+    const MfClassicData* data = nfc_device_get_data(device, NfcProtocolMfClassic);
+
+    do {
+        // verify all sectors' keys
+        for(size_t i = 0; i < mf_classic_get_total_sectors_num(data->type); i++) {
+            MfClassicSectorTrailer* sec_tr = mf_classic_get_sector_trailer_by_sector(data, i);
+            uint64_t key = bit_lib_bytes_to_num_be(sec_tr->key_a.data, 6);
+            if(key != bip_keys_a[i]) {
+                parsed = false;
+                break;
+            }
+            key = bit_lib_bytes_to_num_be(sec_tr->key_b.data, 6);
+            if(key != bip_keys_b[i]) {
+                parsed = false;
+                break;
+            }
+        }
```

Comment:
Is it really necessary to verify all the keys? The execution of this code is quite long and it leads to a little "freeze" before we can see parsed data.

Reply:
I would just check the key that we use in `verify()` function, if it's OK for you

Reply:
I suppose if we already read the card, then the other keys should match. The only unlikely edge case I can think of is loading a stored card where the first sector keys' match but not the others, though that seems unlikely enough.